### PR TITLE
Proposal to update logic behind scenario/map  creation and editing

### PR DESCRIPTION
**Pull request by alkorolyov**
Hi @yairm210 ,

I have tested basic scenario functionality and it seems to work correctly, at least for this last implementation:

- saving
- loading
- creating/editing
- new game start

I was thinking about further separating two ideas: scenario and map. Make them closer to the original Civ 5:

- Map files store only "geographical" information: terrain types and resources.
- Scenario files store maps and additional information regarding civilizations and player: starting locations, chosen civ, units, improvements, wining conditions.

I see it that way, in Map Editor when we first create new map, we have only `Terrains & Resources` button. And when we click "Create scenario", then we can choose number of players, civilizations and game parameters on `GameParameterScreen`. When we return to map editor `Improvements` and `Units` buttons are now visible and we can add units/starting locations/improvements, based on the chosen civs.

We can even strip unit/civilization information when user is saving scenario as a map, or loads map with this additional info.

What do you think about it?

### CODE REVIEW

**Comment by alkorolyov** [comment]
@yairm210 
Could you please check this conflict. I am still not very confident with git hub, especially when it comes to merge confilicts.

As for the work done:
I think the main part is finished. The only important thing is to implement somehow "Random nation", to be able to set starting locations and units for players with Random civ. All functionality is still hidden behind switch.

**Comment by alkorolyov** [comment]
Ok, I was wrong about "Random Nation". In the world builder there is no possibility to create scenario with random nation and add units/starting location for them. That's make it much simpler to implement.

**Comment by alkorolyov** [comment]
@yairm210 
All the basic functionality was implemented and working for the time being.
I can say it was a pleasure to work on this Scenario feature, it was very hard and scary to write first lines of code but then it went smoother :) However I feel that my code style is far from perfect and quite ugly sometimes.
Thank you for letting me contribute to your project!

When you will have time please review these commits and give me some feedback, would be much appreciated. Maybe we can even have a short voice discussion as for me it is sometimes difficult to get complex ideas just from text conversation.

Best regards :)

**Comment by alkorolyov** [comment]
Finally I overcame my autistic behaviour and dive into discord developers/bug discussions and understood the meaning of your commit [`1b73aab`](https://github.com/alkorolyov/Unciv/commit/1b73aab856696e0b3bc83bdbe01493c697615255)
Slightly changed my implementation. And now it is compatible with this feature:
Opening the New Game screen from within a game saves the previous game parameters, map generation parameters work again
Now one of the conflicts in `core/src/com/unciv/ui/newgamescreen/NewGameScreen.kt` is fixed.

**Comment by alkorolyov** [comment]
Finally I overcame my autistic behaviour and dive into discord developers/bug discussions and understood the meaning of your commit [1b73aab](https://github.com/alkorolyov/Unciv/commit/1b73aab856696e0b3bc83bdbe01493c697615255). I have slightly changed my implementation and now it is compatible with this feature:
Opening the New Game screen from within a game saves the previous game parameters, map generation parameters work again

Now one of the conflicts in core/src/com/unciv/ui/newgamescreen/NewGameScreen.kt is fixed.
As for the second conflict in `core/src/com/unciv/ui/mapeditor/GameParametersScreen.kt`:
we need to call both `PlayerPickerTable `from `GameParametersScreen` and not `mapEditorScreen`, because I am using `GameParametersScreen` as temporary variable for `gameSetupInfo` to be able to cancel changes onClose or apply them to `mapEditorScreen` upon confirmation, "OK" click.

**Comment by yairm210** [comment]
About unit clearing, I agree, and I think this should happen when changing the player info.

**Comment by alkorolyov** [comment]
Could you take a look on merge conflict?

File: core/src/com/unciv/logic/GameStarter.kt [hunk 1]
```diff
@@ -120,6 +120,21 @@ object GameStarter {
                 tile.improvement = null
         }
 
+        // set max starting movement for units loaded from map
+        for (tile in gameInfo.tileMap.values) {
+            tile.apply {
```
Comment (yairm210) [inline core/src/com/unciv/logic/GameStarter.kt:125]: for(unit in tile.getUnits()) unit.currentMovement = unit.getMaxMovement()

**Comment by yairm210** [comment]
The commits themselves are fine, it'll all get squashed anyway, but there's TOO MUCH STUFF HERE!
Please split this into smaller PRs =)

File: core/src/com/unciv/logic/map/TileMap.kt [hunk 2]
```diff
@@ -215,6 +217,54 @@ class TileMap {
...
+     * @param player [Player] whose all units and locations will be changed
+     * @param newNation new nation to be set up
+     */
+    fun switchPlayerNation(player: Player, newNation: Nation) {
+        tileList.forEach { it.apply {
+            if (improvement == "StartingLocation " + player.chosenCiv) {
+                improvement = "StartingLocation " + newNation.name
+            }
+
+            if (militaryUnit !=null && militaryUnit?.owner == player.chosenCiv) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/TileMap.kt:253]: Same as above use tile.getUnits()

**Comment by alkorolyov** [comment]
> Please split this into smaller PRs
I think it's a good idea because I messed up the last merging process and really don't know how to bring everything into order.

**Comment by alkorolyov** [comment]
> but there's TOO MUCH STUFF HERE!

Sorry I got too exited and impatient that actually forgot that someone would need to reed all those commits =)

**Pull request closed without merging**